src/fileops.rs

```diff
@@ -1,27 +1,54 @@
 //! Contains operations related to the /dev/ipath character files.
-//! fcntl is not included because rust does not allow for the C-style
-//! argument list (...) outside of ffi functions. fcntl is only used once
-//! in the C version.
 
 extern crate libc;
 
 use std::ffi::CString;
 use libc::c_int;
+use std::os::unix::io::{RawFd, AsRawFd};
+use std::ops::Drop;
 
-pub fn open<T: Into<String>>(path: T, mode: c_int) -> c_int {
-  unsafe { libc::open(CString::new(path.into())
-                      .unwrap_or(CString::new("").unwrap()).as_ptr(), mode) }
+pub struct Fd(RawFd);
+
+impl Fd {
+  pub fn open<T: Into<String>>(path: T, mode: c_int) -> Option<Fd> {
+    let fd = unsafe { libc::open(CString::new(path.into())
+                .unwrap_or(CString::new("").unwrap()).as_ptr(), mode) };
+    match fd {
+      -1 => None,
```

Comment:
Dumb comment i think: It looks like this pattern is really common. Is it possible to raise this like so:

```Rust
// Contract: Fd != -1
struct SafeDescriptor(Fd);

impl SafeDescriptor {
  unsafe fn get_mut_descriptor(&self) -> ...
  fn get_descriptor(&self) -> ...
  fn from_descriptor(descriptor: Fd) -> Option<SafeDescriptor> {
    if descriptor == -1 {
      None
    }  else {
      Some(SafeDescriptor(descriptor))
    }
  }
}
```

It adds extra boilerplate everwhere, but it codifies the contract you're trying to make -- namely that you have a valid FD. Thoughts?